### DISCUSSION TITLE
當碰到併發(Concurrent)請求時，NodeJs服務器會怎麼執行呢？

### DISCUSSION THREAD

**Discussion by TehoChang** [Lidemy/forum#256] [後端相關問題]
這幾天需要寫一個序列號產生器API，邏輯很簡單，每發送一個http request 給API，API返回一個連續序列號。簡單的pseudo code如下：
1 讀取資料庫中記錄的最後一個號碼，定義為lastNumber
2 lastNumber +1，返回給使用者

然後，有了關於併發的疑問，如果同時有大量的請求發送到序列號產生器會發生什麼事呢？ 之後我又產生更多關於nodeJs server的問題，發現自己的基礎知識不行啊！想請教的問題如下：
**一、 nodeJs是單線程語言，一次只會處理一個請求，正常情況，產生器應該依序產生001, 002, 003...，那麼會不會出現，同時有大量的請求發送到產生器API，這些請求從資料庫讀取到的都是同一個最後序列號，因而返回重複的序列號的情況呢？ 舉例來說，好幾個請求同時讀取到 lastNumber=001，因此都返回002**

陸續看了很多文章，這一篇蠻清楚的  https://stackoverflow.com/questions/34855352/how-in-general-does-node-js-handle-10-000-concurrent-requests
其中有畫出nodeJs 處理請求的模型
- 單個請求的情況
user do an action
            │
            v
  application start processing action
   └──> make database request
          └──> do nothing until request completes
  request complete
   └──> send result to user

- 多個請求的情況
 request A ──> make database request
 request B ──> make database request
 request C ──> make database request
database request complete ──> send response A
database request complete ──> send response B
database request complete ──> send response C

那麼疑問來了，怎樣算一個DB request complete呢？ 產生器程式對於資料庫的操作是一讀一寫，讀取LastNumber，寫入LastNumber+1。 是兩個動作都執行完，算DB request complete嗎？
而又由於Node是單線程，資料庫處理資料時，node不會閒著，會接下去處理request B, request C，那針對request B, request C執行的make DB request，又會進行到哪一階段呢？ 請參考下圖，粗體數字是我想像中的動作完成順序
 **1** request A ──> make database request ––>  **4 DB處理資料(read 001) (write 002)** 
**2** request B ──> make database request ––>??（**疑問所在： 執行順序？ 會執行到哪階段？**）
 **3** request C ──> make database request
**5** database request complete ──> send response A
database request complete ──> send response B
database request complete ──> send response C
 
 當**4**沒有完成的時候，request B的make DB request會執行到哪裡呢？ 因為lastNumber要完成更新到002，request B才能成功返回003。
 Q: DB本身的程式似乎是多線程？

資料庫處理資料有一種機制叫事務（transaction)，不知道是不是我疑問的解答？

**二、 假設有一個nodeJS REST API中有許多routes，除了產生器route，還有其他很多routes負責不同服務，那麼，是每個route算一個單線程單位嗎？還是一整個REST API Server算是一個單線程單位呢？
舉例來說：序列號route處理很多請求的同時，其他routes可以同時處理各自的API request嗎？ 還是所有routes要按照順序一個一個排隊呢？**

還望大家可以多多指教，問題可能整理得不太清楚，需要釐清的地方歡迎提出，我再加強解釋，感謝～

**Comment by cwc329** [comment]
看起來你的序列號需要依照 DB 裡面的資料來決定，如果要確保 a, b , c 三個 request 能夠有照順序的標號，依照我的理解這樣就與 nodejs 沒有大關係，這邊要注意的反而是 DB 的 race condition。
DB通常都會有 lock 機制，來確避免 race condition。而且我記得以 MySQL 來說要使用 lock 是要用在 transaction 裡面的。
不知道有沒有幫到你，如果有理解錯誤請不吝告知。

**Comment by aszx87410** [comment]
Nodejs 的執行就跟在瀏覽器上差不多，核心都是 event loop，你只要有一個同步的 request 卡住，後續全部都會卡住，所以通常 query DB 之類的都是非同步，在非同步的時候就可以先去處理其他 request，等資料回來時靠著 event loop 再繼續執行

不過同樓上所說，你這問題跟 nodejs 本身沒太大關係，想確保產生序列號的話要用到 DB 的 lock 跟 transaction

**Reply by TehoChang** [reply]
請問關於問題二，應該用什麼觀念去理解呢？
我閱讀過您event loop的文章，也看過那個youtube影片，還是不知道問題二的正解是什麼？ 該如何理解?

謝謝老師~

**Reply by aszx87410** [reply]
server 其實就只是執行一個 JS 檔案而已，所以用一般理解 JS 方式就好，一整個 server 都是 single thread，所以如果這樣寫：

``` js
const express = require('express')
const app = express()
const port = 3000

app.get('/wait', (req, res) => {
  while(1) {}
})

app.get('/', (req, res) => {
  res.send('Hello World!')
})

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})
```

當你拜訪 /wait 之後，拜訪 / 也會卡住，因為 JS 被卡在 /wait 那邊了

**Reply by TehoChang** [reply]
謝謝Huli~ 碰到問題時我應該自己有能力寫這種簡單案例來測試！
